### ISSUE TITLE
The return type of waitForSelector is Promise<ElementHandle | null> even when hidden is false

### ISSUE THREAD

**Issue by redneb** [puppeteer/puppeteer#7295]
With the new built-in TypeScript definitions, the type of the `waitForSelector` method is

```
waitForSelector(selector: string, options?: WaitForSelectorOptions): Promise<ElementHandle | null>;
```

which means that the type of the returned value is `Promise<ElementHandle | null>`. But according to the documentation, if the `hidden` option is set to `false` (which is the default), then `waitForSelector` is always going to return a non-`null` value (or throw an exception). By unconditionally including `null` in the return type, users have to either perform a check for `null`ness at runtime, or use a type assertion which reduces type safety. Both solutions are not ideal.

**Comment by casey6** [comment]
Relevant `@types/puppeteer` type definitions: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/puppeteer/index.d.ts#L1331-L1332

**Comment by SimonSchick** [comment]
Bump.

**Comment by jespertheend** [comment]
This should be fixed with #7942

**Comment by jrandolf-zz** [comment]
Based on @jespertheend's PR, this type signature is not something we'd like to support. In particular, someone could manually write true and since hidden can be undefined, the return type would be lying to the developer. And even if we fixed this, the type signature would almost never be used since an options object will generally genericize the boolean attribute of hidden, so when passed into the function, it would go back to using boolean. Unless the options object manually types `true` for the `hidden` attribute, this type is meaningless.

**Comment by CodyEakins** [comment]
@jrandolf-zz 

I want to revisit this issue.
You closed @jespertheend's PR (#7942) because they used generics to solve the problem. I agree with you, in that generics can sometimes be unreliable and awkward to use, but what about using overloads instead?

Here's an example from DefinitelyTyped's Puppeteer definitions:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/puppeteer/index.d.ts#L1341-L1342

If you or the team will allow this type of change, I'll write the PR to implement it on all the relevant `waitFor...` methods.
Just looking for the "okay" before wasting time.
